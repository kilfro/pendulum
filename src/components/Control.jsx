import React from 'react'
import '@style/control.less'

const Control = () => {
    return (
        <div className="control">
            <div className="control-group">
                <label htmlFor="thread-length">Длина подвеса</label>
                <input type="range" name="thread-length"/>
            </div>
            <div className="control-group">
                <label htmlFor="ball-weight">Вес груза</label>
                <input type="range" name="ball-weight"/>
            </div>
            <div className="control-group">
                <label htmlFor="gravity">Сила притяжения</label>
                <input type="range" name="gravity"/>
            </div>
            <div className="control-group">
                <label htmlFor="momentum">Начальный импульс</label>
                <input type="range" name="momentum"/>
            </div>
        </div>
    )
}

export default Control