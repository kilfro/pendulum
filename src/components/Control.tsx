import React from 'react'
import {IControlMap} from '../types/ContorolTypes'
import {useTypedSelector} from '../hooks/useTypedSelector'
import {useActions} from '../hooks/useActions'
import styled from 'styled-components'

const ControlSection = styled.div`
  background-color: rgba(95, 135, 120, 0.6);
  box-sizing: border-box;
  padding-top: 20px;
  display: flex;
  flex-direction: column;
  align-items: center;
`

const ControlGroup = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;
  align-items: center;
  justify-content: space-between;
  font-size: 18px;
  width: 500px;
  margin-bottom: 20px;

  label {
    text-transform: uppercase;
    color: white;
  }

  input {
    width: 300px;
  }
`

export const Control: React.FC = () => {
    const {length, impulse, gravity, weight} = useTypedSelector(state => state)
    const {setLength, setGravity, setWeight, setImpulse} = useActions()

    const settersMap: IControlMap = {
        length: {
            description: 'Длина подвеса',
            min: 10,
            max: 30,
            step: 0.1,
            value: length,
            set: setLength,
        },
        weight: {
            description: 'Вес груза',
            min: 20,
            max: 100,
            value: weight,
            set: setWeight,
        },
        gravity: {
            description: 'Сила притяжения',
            min: 1,
            max: 100,
            step: 0.1,
            value: gravity,
            set: setGravity,
        },
        impulse: {
            description: 'Начальный импульс',
            min: 1,
            max: 60,
            value: impulse,
            set: setImpulse,
        },
    }

    return (
        <ControlSection>
            {
                Object.entries(settersMap).map(([key, option]) => {
                        const {set, description, ...rest} = option
                        const inputProps = {
                            ...rest,
                            name: key,
                            onChange: (event: React.ChangeEvent<HTMLInputElement>) => {
                                set(event.target.value)
                            },
                        }

                        return (
                            <ControlGroup key={key}>
                                <label htmlFor={key}>{description}: {option.value}</label>
                                <input type="range" {...inputProps}/>
                            </ControlGroup>)
                    },
                )
            }
        </ControlSection>
    )
}