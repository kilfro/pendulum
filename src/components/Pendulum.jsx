import React, {useEffect, useState} from 'react'
import '@style/pendulum.less'
import {connect} from 'react-redux'

export const Pendulum = ({length, impulse, gravity, weight}) => {
    const [time, setTime] = useState(0)
    const [angel, setAngel] = useState(0)
    const timeout = 5

    const calculateCurrentAngel = (timeInSeconds) => {
        return impulse * Math.sin(Math.sqrt(gravity / (length / 100)) * timeInSeconds)
    }

    useEffect(() => {
        const intervalId = setInterval(() => {
            setAngel(calculateCurrentAngel(time / 1000))
            setTime(prev => prev + timeout)
        }, timeout)

        return () => clearInterval(intervalId)
    }, [time])

    const pendulumStyle = {
        height: `${length}cm`,
        transform: `rotate(${angel}deg)`
    }

    const ballStyle = {
        bottom: -weight,
        left: weight * -0.5,
        width: `${weight}px`,
        height: `${weight}px`,
    }

    return (
        <div className="pendulum">
            <div className="thread" style={pendulumStyle}>
                <div className="ball" style={ballStyle}/>
            </div>
        </div>
    )
}
const mapStateToProps = state => state

export default connect(mapStateToProps)(Pendulum)