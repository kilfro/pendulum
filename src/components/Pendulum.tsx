import React, {useEffect, useState} from 'react'
import {useTypedSelector} from '../hooks/useTypedSelector'
import styled from 'styled-components'
import {BallProps, ThreadProps} from '../types/PendulumTypes'

const PendulumSection = styled.div`
  margin: 0 auto;
`

const Thread = styled.div<ThreadProps>`
  position: relative;
  border: 1px solid steelblue;
  transform-origin: top;
  height: ${props => props.length + 'cm'};
  transform: ${props => 'rotate(' + props.angel + 'deg)'};
`

const Ball = styled.div<BallProps>`
  position: absolute;
  background-color: rgba(75, 0, 130, 0.96);
  border-radius: 50%;
  bottom: ${props => -props.weight + 'px'};
  left: ${props => +props.weight * -0.5 + 'px'};
  width: ${props => props.weight + 'px'};
  height: ${props => props.weight + 'px'};
`

export const Pendulum: React.FC = () => {
    const timeout = 5

    const [time, setTime] = useState<number>(0)
    const [angel, setAngel] = useState<number>(0)
    const {weight, impulse, gravity, length} = useTypedSelector(state => state)

    const calculateCurrentAngel = (timeInSeconds: number): number => {
        return +impulse * Math.sin(Math.sqrt(+gravity / (+length / 100)) * timeInSeconds)
    }

    useEffect(() => {
        const intervalId = setInterval(() => {
            setAngel(calculateCurrentAngel(time / 1000))
            setTime(prev => prev + timeout)
        }, timeout)

        return () => clearInterval(intervalId)
    }, [time])

    return (
        <PendulumSection>
            <Thread length={length} angel={angel}>
                <Ball weight={weight}/>
            </Thread>
        </PendulumSection>
    )
}