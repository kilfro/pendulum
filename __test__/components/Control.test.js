import React from 'react'
import {Control} from '../../src/components/Control'
import {shallow} from 'enzyme'

describe(Control, () => {
    test('should match snapshot', () => {
        const component = shallow(
            <Control
                weight="10"
                length="100"
                gravity="10"
                impulse="30"
            />
        )

        expect(component).toMatchSnapshot()
    })

    test('should call dispatch functions', () => {
        const setLengthMock = jest.fn()
        const setWeightMock = jest.fn()
        const setGravityMock = jest.fn()
        const setImpulseMock = jest.fn()

        const component = shallow(
            <Control
                weight="10"
                length="100"
                gravity="10"
                impulse="30"
                setThreadLength={setLengthMock}
                setWeight={setWeightMock}
                setGravity={setGravityMock}
                setInitialImpulse={setImpulseMock}
            />
        )

        component.find('input[name="length"]').simulate('change', {target: {value: '50'}})
        expect(setLengthMock).toHaveBeenCalledWith('50')

        component.find('input[name="weight"]').simulate('change', {target: {value: '45'}})
        expect(setWeightMock).toHaveBeenCalledWith('45')

        component.find('input[name="gravity"]').simulate('change', {target: {value: '5'}})
        expect(setGravityMock).toHaveBeenCalledWith('5')

        component.find('input[name="impulse"]').simulate('change', {target: {value: '10'}})
        expect(setImpulseMock).toHaveBeenCalledWith('10')
    })
})