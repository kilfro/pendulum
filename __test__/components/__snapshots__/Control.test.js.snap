// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Control should match snapshot 1`] = `ShallowWrapper {}`;
